sale_coupon_advanced/models/sale_coupon.py

```diff
@@ -0,0 +1,38 @@
+import secrets
+import string
+
+from odoo import _, api, fields, models
+from odoo.exceptions import UserError
+
+
+class SaleCoupon(models.Model):
+    """Extend to implement custom code generation rule."""
+
+    _inherit = "sale.coupon"
+
+    def _validate_custom_code_rule(self, custom_code_rule):
+        if custom_code_rule[1] < 1:
+            raise UserError(
+                _(
+                    "Custom Code Rule Must have at least one random "
+                    "alphanumeric symbol"
+                )
+            )
+
+    def _generate_custom_code(self, custom_code_rule):
+        alphabet = string.ascii_letters + string.digits
+        prefix, nbr = custom_code_rule
+        code = "".join(secrets.choice(alphabet) for i in range(nbr))
+        return "{}{}".format(prefix, code)
+
+    @api.model
+    def _generate_code(self):
+        """Override to handle custom_code_rule from context."""
+        custom_code_rule = self._context.get("custom_code_rule")
+        if custom_code_rule:
+            self._validate_custom_code_rule(custom_code_rule)
+            return self._generate_custom_code(custom_code_rule)
```

Comment:
AFAIS the client may be got to the state when generated code already in the use, like when prefix and amount of numbers are too short,
 maybe we need a check here if the code is unique and request a new one if not, otherwise it will fail with sql constrains

Reply:
But thats what constraint will do anyway, and will prevent from generating that coupon. Wouldn't this be redundant if we would check same thing twice?

Reply:
ok